File: Remove-Nth-Node-From-End-of-List/Remove-Nth-Node-From-End-of-List.cpp

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        if(!head)
        {
            return head;
        }
        ListNode *p=head;
        ListNode *q=head;
        while(n--)
        {
            q=q->next;
        }
        if(!q)
            head=head->next;
        else
        {
            while(q->next)
            {
                p=p->next;
                q=q->next;
            }
            p->next=p->next->next;
        }
        return head;
    }
};


 

File: solutions/Remove-Nth-Node-From-End-of-List/Remove-Nth-Node-From-End-of-List.py

‘’’
 Definition for singly-linked list.
 class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None
‘’’
class Solution:
    # @return a ListNode
    def removeNthFromEnd(self, head, n):
        p=head
        q=head
        s=None
        for i in range(1,n):
            q=q.next
        while q.next!=None:
            s=p
            p=p.next
            q=q.next
        if s==None:
            head=p.next
        else:
            s.next=p.next
        return head


 